jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DosHandler.java

```diff
@@ -0,0 +1,559 @@
+                return !isRateExceeded(now, false) && _exponentialMovingAverage <= 0.0001;
+            }
+
+            double getCurrentRatePerSecond()
+            {
+                return _exponentialMovingAverage;
+            }
+        }
+    }
+
+    /**
+     * A Handler to reject DoS requests with a status code or failure.
+     */
+    public static class StatusRejectHandler implements Request.Handler
+    {
+        private final int _status;
+
+        public StatusRejectHandler()
+        {
+            this(-1);
+        }
+
+        /**
+         * @param status The status used to reject a request, or 0 to fail the request or -1 for a default ({@link HttpStatus#TOO_MANY_REQUESTS_429}.
+         */
+        public StatusRejectHandler(int status)
+        {
+            _status = status >= 0 ? status : HttpStatus.TOO_MANY_REQUESTS_429;
+        }
+
+        @Override
+        public boolean handle(Request request, Response response, Callback callback) throws Exception
+        {
+            if (_status == 0)
+                callback.failed(new RejectedExecutionException());
+            else
+                Response.writeError(request, response, callback, _status);
+            return true;
+        }
+    }
+
+    /**
+     * A Handler to reject DoS requests after first delaying them.
+     */
+    public static class DelayedRejectHandler extends Handler.Abstract
+    {
+        record Exchange(Request request, Response response, Callback callback)
+        {}
+
+        private final AutoLock _lock = new AutoLock();
+        private final Deque<Exchange> _delayQueue = new ArrayDeque<>();
+        private final int _maxDelayQueue;
+        private final long _delayMs;
+        private final Request.Handler _reject;
+        private Scheduler _scheduler;
+
+        public DelayedRejectHandler()
+        {
+            this(-1, -1, null);
+        }
+
+        /**
+         * @param delayMs The delay in milliseconds to hold rejected requests before sending a response or -1 for a default (1000ms)
+         * @param maxDelayQueue The maximum number of delayed requests to hold or -1 for a default (1000ms).
+         * @param reject The {@link Request.Handler} used to reject {@link Request}s or null for a default ({@link HttpStatus#TOO_MANY_REQUESTS_429}).
+         */
+        public DelayedRejectHandler(
+            @Name("delayMs") long delayMs,
+            @Name("maxDelayQueue") int maxDelayQueue,
+            @Name("reject") Request.Handler reject)
+        {
+            _delayMs = delayMs >= 0 ? delayMs : 1000;
+            _maxDelayQueue = maxDelayQueue >= 0 ? maxDelayQueue : 1000;
+            _reject = Objects.requireNonNullElseGet(reject, () -> new StatusRejectHandler(HttpStatus.TOO_MANY_REQUESTS_429));
+        }
+
+        @Override
+        protected void doStart() throws Exception
+        {
+            super.doStart();
+            _scheduler = getServer().getScheduler();
+            addBean(_scheduler);
```

Comment:
I would `installBean()` in the constructor to avoid the hassle of NPE when stopping but requests are still flowing.

Reply:
Server is not set in the constructor.  Best I can do is move to an override of `setServer`, but I think it is cleaner to add/remove where it is created/nulled